### DISCUSSION TITLE
Static pages and A/B testing with middleware - HOW?

### DISCUSSION THREAD

**Discussion by zmarty** [vercel/next.js#33511] [Help] [answered]
Our goal: Have two **static** pages index.tsx and index2.tsx used in an A/B experiment where we use middleware rewrite() to switch between them.

I want to mention that we host Next.js in Azure on our own, and we do not use Vercel.

The problem we are having is that for static pages Next.js emits a cache control header that makes our CDN cache one page or another. And we cannot override cache control headers (to set it to no-cache) for static pages in next.config.js. Here's what the [docs](https://nextjs.org/docs/going-to-production#caching) say: _Cache-Control headers set in next.config.js will be overwritten in production to ensure that static assets can be cached effectively._ 

It's probably something very simple that I am missing, but I don't understand how we can have two static pages (built at Next.js build time with GetStaticPageProps) involved in an A/B test directed by middleware, and also use a CDN in front?

**Comment by leerob** [comment, marked as answer]
Wrote some thoughts and included a few examples here: https://vercel.com/docs/concepts/solutions/ab-testing

**Reply by zmarty** [reply]
@leerob Sorry I don't really get which part specifically fixes our problem :) So how can I A/B test two **static** pages on Next.js hosted **outside** of Vercel?

**Reply by leerob** [reply]
See the examples link, these Next.js applications all work self-hosted too: https://github.com/vercel/examples/tree/main/edge-functions/ab-testing-simple

**Reply by zmarty** [reply]
Hello again. I am not questioning if A/B testing works when self-hosted, it does. The problem is HOW it works in terms of caching. I opened a bug and it was also converted into a discussion. You can see my repro code there: https://github.com/vercel/next.js/discussions/33670

**Reply by eerikson** [reply]
Hi! Was there any resolution here? Seems like an answer was chosen by someone other than OP, who didn't really get what they were looking for. I'm having the same question.
